### PR TITLE
conditional enhancer

### PR DESCRIPTION
**Pull request by dbu**
alternative to #172 

@ElectricMaxxx this is what i had in mind. it would keep the changes minimal and have no impact for people not needing conditional enhancers.

### CODE REVIEW

**Comment by ElectricMaxxx** [comment]
That looks really nice, but i miss some points:

* How do you get that enhancerMap?
* Where would you pass the the mapping informormation, that tree of five current enhancers would need? Still through service defintion or the constructor
* The matcher just decides for the responsibility of the enhancer? So it would be aware of all "early outs" we currently have in the enhancers? Why can't the enhancer have a `match()` method?
* Where would you see the point of preparing mapping information like
```
[
    'methods' => ['put', 'post'],
   'value' => 'service:postAndPutAction'
]
```
into
```
[
    'service:postAndPutAction'
]
```
for POST/PUT request and nothing for all other requests? Caus the `['service:postAndPutAction']` is the only mapping the `FieldByClass` enhancer (with target=_controller)  is able to work with, to get the right controllerAction for the resolver.

**Comment by dbu** [comment]
>   * How do you get that enhancerMap?

for me, the configuration is the job of the bundle, not of the library.
the library works with object graphs.

>   * Where would you pass the the mapping informormation, that tree of
>     five current enhancers would need? Still through service defintion
>     or the constructor

service definitions that are created from the configuration.

>   * The matcher just decides for the responsibility of the enhancer? So
>     it would be aware of all "early outs" we currently have in the
>     enhancers? Why can't the enhancer have a |match()| method?

i don't want to complicate the enhancers. the matcher is used to
determine if a specific enhancer is applicable for this request. the
enhancers should not need to know about that, they would be duplicating
the request matchers.

>   * Where would you see the point of preparing mapping information like
> 
> |[ 'methods' => ['put', 'post'], 'value' => 'service:postAndPutAction' ] |
> 
> into
> 
> |[ 'service:postAndPutAction' ] |
> 
> for POST/PUT request and nothing for all other requests? Caus the
> |['service:postAndPutAction']| is the only mapping the |FieldByClass|
> enhancer (with target=_controller) is able to work with, to get the
> right controllerAction for the resolver.

```php
$controllerEnhancer = new FieldByClassEnhancer(
    '_content',
    '_controller',
    ['Symfony\Cmf\Component\Routing\Tests\Enhancer\RouteObject' =>
'cmf_content.controller:postAndPutAction']
);
$putPostMatcher = new RequestMatcher(null, null, ['put', 'post']);

$conditionalEnhancer = new ConditionalEnhancer([
    ['matcher' => $putPostMatcher, 'enhancer' => $controllerEnhancer],
]);

```

because we add no further enhancer, this thing will only act on post/put
requests.

**Comment by dbu** [comment]
fixed the bugs i had in this draft. i think its now ready to merge.

File: Tests/Enhancer/ConditionalEnhancerTest.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            ->method('enhance')
+            ->with($defaults, $this->request)
+            ->will($this->returnValue($expected));
+        $matcher2 = $this->buildMock('\Symfony\Component\HttpFoundation\RequestMatcherInterface');
+        $matcher2->expects($this->once())
+            ->method('matches')
+            ->with($this->request)
+            ->will($this->returnValue(true));
+
+
```
Comment (ElectricMaxxx) [inline Tests/Enhancer/ConditionalEnhancerTest.php]: Can you please remove the second empty line to satisfy StyleCI and me.

Comment (dbu) [inline Tests/Enhancer/ConditionalEnhancerTest.php]: fixed. i am now off until next wednesday

**Comment by ElectricMaxxx** [comment]
thanks a lot for your effort to explain and solving the problem.

**Pull request merged**